apps/aecore/src/aec_db_gc.erl

```diff
@@ -0,0 +1,206 @@
+handle_event(info, {_, top_changed, #{info := #{height := Height}}}, idle,
+             #data{interval = Interval, history = History,
+                   enabled = true, synced = true,
+                   height = undefined, hashes = undefined} = Data)
+  when Height rem Interval == 0 ->
+    Top = self(),
+    Pid = spawn_link(
+            fun () ->
+                    FromHeight = max(Height - History, 0),
+                    {Time, {ok, Hashes}} = ?TIMED(collect_reachable_hashes(FromHeight, Height)),
+                    ets:give_away(Hashes, Top, Time)
+            end),
+    {keep_state, Data#data{height = Height, hashes = Pid}};
+handle_event(info, {'ETS-TRANSFER', Hashes, _, Time}, idle, #data{enabled = true, hashes = Pid} = Data)
+  when is_pid(Pid) ->
+    ?LOG("GC scanning of ~p reachable hashes took ~p seconds",
+         [ets:info(Hashes, size), Time / 1000000]),
+    {next_state, ready, Data#data{hashes = Hashes}};
+
+handle_event(info, {_, top_changed, #{info := #{height := Height}}}, ready,
+             #data{enabled = true, synced = true, height = LastHeight, hashes = Hashes} = Data)
+  when is_reference(Hashes), Height > LastHeight ->
+    {ok, _} = range_collect_reachable_hashes(Height, Data),
+    {keep_state, Data#data{height = Height}};
+
+handle_event({call, From}, maybe_garbage_collect, ready,
+             #data{enabled = true, synced = true, hashes = Hashes} = Data)
+  when Hashes /= undefined, not is_pid(Hashes) ->
+    Header = aec_chain:top_header(),
+    case aec_headers:type(Header) of
+        key ->
+            Height  = aec_headers:height(Header),
+            {ok, _} = range_collect_reachable_hashes(Height, Data),
+            {ok, N} = swap_nodes(Hashes),
+            ets:delete(Hashes),
+            {next_state, idle, Data#data{height = undefined, hashes = undefined},
+             {reply, From, {ok, N}}};
+        micro ->
+            {keep_state, Data}
+    end;
+handle_event({call, From}, maybe_garbage_collect, _, Data) ->
+    {keep_state, Data, {reply, From, nop}};
+
+handle_event(_, _, _, Data) ->
+    {keep_state, Data}.
+
+
+terminate(_Reason, _State, _Data) -> void.
+
+code_change(_, State, Data, _) -> {ok, State, Data}.
+
+callback_mode() -> handle_event_function.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% From - To (inclusive)
+collect_reachable_hashes(FromHeight, ToHeight) when FromHeight < ToHeight ->
+    {ok, Hashes} = collect_reachable_hashes_fullscan(FromHeight),
+    {ok, Hashes} = range_collect_reachable_hashes(FromHeight, ToHeight, Hashes),
+    {ok, Hashes}.
+
+collect_reachable_hashes_fullscan(Height) ->
+    Tab = ets:new(gc_reachable_hashes, [public]), %% TODO: remove public
+    MPT = get_mpt(Height),
+    {ok, aeu_mp_trees:visit_reachable_hashes(MPT, Tab, fun store_hash/3)}.
+
+%% assumes Height - 1, Height - 2, ... down to Height - History
```

Comment:
If the chain reorganized between the `Height` and `Height - History`, it emits top_change as well for middle blocks? (if not we can miss some blocks in `hashes`)